source/common/common/utility.cc

```diff
@@ -20,8 +22,72 @@
 #include "spdlog/spdlog.h"
 
 namespace Envoy {
+
+namespace {
+
+class SubsecondConstantValues {
+public:
+  const size_t DEFAULT_SECONDS_LENGTH{10};
+  const char PLACEHOLDER_CHAR{'?'};
+  const std::string PLACEHOLDER{"?????????"};
+  const std::regex PATTERN{"%([1-9])?f", std::regex::optimize};
+};
+
+typedef ConstSingleton<SubsecondConstantValues> SubsecondConstants;
+
+} // namespace
+
 std::string DateFormatter::fromTime(const SystemTime& time) const {
-  return fromTime(std::chrono::system_clock::to_time_t(time));
+  struct CachedTime {
+    std::chrono::seconds epoch_time_seconds;
+    std::unordered_map<std::string, const std::string> formatted;
+  };
+  static thread_local CachedTime cached_time;
+
+  const std::chrono::nanoseconds epoch_time_ns =
+      std::chrono::duration_cast<std::chrono::nanoseconds>(time.time_since_epoch());
+
+  const std::chrono::seconds epoch_time_seconds =
+      std::chrono::duration_cast<std::chrono::seconds>(epoch_time_ns);
+
+  if (cached_time.formatted.find(format_string_) == cached_time.formatted.end() ||
+      cached_time.epoch_time_seconds != epoch_time_seconds) {
+    cached_time.formatted.emplace(
+        std::make_pair(format_string_, fromTime(std::chrono::system_clock::to_time_t(time))));
+
+    cached_time.epoch_time_seconds = epoch_time_seconds;
+  }
+
+  if (subsecond_widths_.empty()) {
+    return cached_time.formatted.at(format_string_);
+  }
+
+  // Copy the cached formatted string, then replace its subseconds part.
+  std::string formatted = cached_time.formatted.at(format_string_);
+  const std::string value = fmt::FormatInt(epoch_time_ns.count()).str();
+  for (const auto width : subsecond_widths_) {
+    const std::string subsecond =
+        value.substr(SubsecondConstants::get().DEFAULT_SECONDS_LENGTH, width);
+    const size_t placeholder_index = formatted.find(SubsecondConstants::get().PLACEHOLDER_CHAR);
+    ASSERT(placeholder_index != std::string::npos);
+    formatted.replace(placeholder_index, width, subsecond);
+  }
+
+  return formatted;
+}
+
+std::string DateFormatter::parse(const std::string& format_string) {
+  std::string new_format_string = format_string;
+  std::smatch result;
+  while (regex_search(new_format_string, result, SubsecondConstants::get().PATTERN)) {
+    const std::string& width_specifier = result[1];
+    const size_t width = width_specifier.empty() ? SubsecondConstants::get().PLACEHOLDER.size()
+                                                 : width_specifier.at(0) - '0';
+    subsecond_widths_.emplace_back(width);
+    new_format_string.replace(result.position(), result.length(),
```

Comment:
Yep, if you remember the start index here alongside the width, you can then avoid the `find` on the data path.

Reply:
Updated.